utils/load_path.ml

```diff
@@ -48,32 +48,23 @@ let dirs = s_ref []
 
 let reset () =
   assert (not Config.merlin || Local_store.is_bound ());
-  files := SMap.empty;
-  files_uncap := SMap.empty;
+  STbl.clear !files;
+  STbl.clear !files_uncap;
   dirs := []
 
 let get () = List.rev !dirs
 let get_paths () = List.rev_map Dir.path !dirs
 
-let add_to_maps fn basenames files files_uncap =
-  List.fold_left (fun (files, files_uncap) base ->
-      let fn = fn base in
-      SMap.add base fn files,
-      SMap.add (String.uncapitalize_ascii base) fn files_uncap
-    ) (files, files_uncap) basenames
-
 (* Optimized version of [add] below, for use in [init] and [remove_dir]: since
    we are starting from an empty cache, we can avoid checking whether a unit
    name already exists in the cache simply by adding entries in reverse
    order. *)
 let add dir =
-  assert (not Config.merlin || Local_store.is_bound ());
-  let new_files, new_files_uncap =
-    add_to_maps (Filename.concat dir.Dir.path)
-      dir.Dir.files !files !files_uncap
-  in
-  files := new_files;
-  files_uncap := new_files_uncap
+  List.iter (fun base ->
+      let fn = Filename.concat dir.Dir.path base in
+      STbl.replace !files base fn;
+      STbl.replace !files_uncap (String.uncapitalize_ascii base) fn
+    ) dir.Dir.files
```

Comment:
Is the comment above ("simply by adding entries in reverse order") still correct/useful/relevant? To be honest I am not sure of what it meant when it was introduced by @nojb in #9611. This sounds likes something worth checking somewhat carefully to avoid ordering issues.

Reply:
It is still relevant. It refers to processing the include dirs in reverse order in the `init` function so that earlier entries override later ones, rather than the other way around.

Reply:
After reading the whole file, I get it. The comment says that entries are added in reverse order, but `remove_dir` does not perform any reversal operation. The key point is that directories are stored in !dirs from most recent to oldest, so doing an overwriting-addition traversal on `!dirs` implements the expected oldest-entry-wins semantics.